kong/tools/dns_resolver.lua

```diff
@@ -0,0 +1,81 @@
+local Object = require "classic"
+local stringy = require "stringy"
+local cache = require "kong.tools.database_cache"
+local resolver = require "resty.dns.resolver"
+
+local DEFAULT_TIMEOUT = 2000
+local DEFAULT_ATTEMPTS = 5
+
+local DnsResolver = Object:extend()
+
+function DnsResolver:new(host, port, timeout, attempts)
+  self.host = host
+  self.port = port
+  self.timeout = timeout or DEFAULT_TIMEOUT
+  self.attempts = attempts or DEFAULT_ATTEMPTS
+end
+
+function DnsResolver:query(address, type)
+  -- Init the resolver
+  local r, err = resolver:new{
+    nameservers = {{self.host, self.port}},
+    retrans = self.attempts,
+    timeout = self.timeout
+  }
+  if not r then
+    return nil, "Startup error: "..err
+  end
+
+  -- Make query
+  local answers, err = r:query(address, {qtype = r["TYPE_"..type]})
+  if not answers then
+    return nil, "failed to query the DNS server: ", err
+  end
+
+  if answers.errcode then
+    return nil, "server returned error code: ", answers.errcode, ": ", answers.errstr
+  end
+
+  return answers
+end
+
+function DnsResolver:resolve(address, port, enable_srv)
+  -- Retrieve from cache
+  local cache_key = cache.dns_key(address)
+  local result = cache.get(cache_key)
+  if result then
+    return {host = result.host, port = result.port}
+  end
+
+  -- Query for A record
```

Comment:
See [RFC2782](https://tools.ietf.org/html/rfc2782). But apparently consul has multiple ways of looking up the same service; 'Standard lookup' and 'RFC 2782 lookup', see https://www.consul.io/docs/agent/dns.html

The way I understand it if you define a service in Consul it constructs multiple ways to retrieve the same service info.

It will be known by this name; `[tag.]<service>.service[.datacenter].<domain>`
as well as; `_<service>._<protocol>.service[.datacenter][.domain]`

They return both a port as well as a hostname/ip. So in case of a hostname, a second lookup may be required.

try; `dig srvtest.thijsschreijer.nl SRV` (not Consul, but regular DNS)